File: server/JobTrack.UiTests/Pages/LoginPage.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;

namespace JobTrack.UiTests.Pages;

public class LoginPage
{
    private const string BaseUrl =
        "http://localhost:4200";

    private readonly IWebDriver _driver;
    private readonly WebDriverWait _wait;


    public LoginPage(IWebDriver driver)
    {
        _driver = driver;

        _wait = new WebDriverWait(
            driver,
            TimeSpan.FromSeconds(10)
        );
    }


    private IWebElement EmailInput =>
        _wait.Until(driver =>
            driver.FindElement(
                By.Name("email")
            )
        );


    private IWebElement PasswordInput =>
        _driver.FindElement(
            By.Name("password")
        );


    private IWebElement SignInButton =>
        _driver.FindElement(
            By.CssSelector(
                ".submit-button"
            )
        );


    public void Open()
    {
        _driver.Navigate().GoToUrl(
            $"{BaseUrl}/login"
        );
    }


    public void Login(
        string email,
        string password)
    {
        EmailInput.Clear();
        EmailInput.SendKeys(email);

        PasswordInput.Clear();
        PasswordInput.SendKeys(password);

        SignInButton.Click();
    }


    public bool IsErrorDisplayed()
    {
        var error =
            _wait.Until(driver =>
                driver.FindElement(
                    By.CssSelector(
                        ".error-message"
                    )
                )
            );

        return error.Displayed;
    }


    public string GetErrorMessage()
    {
        return _wait.Until(driver =>
            driver.FindElement(
                By.CssSelector(
                    ".error-message"
                )
            )
        ).Text;
    }


    public void WaitForSuccessfulLogin()
    {
        _wait.Until(driver =>
            !driver.Url.Contains(
                "/login"
            )
        );
    }
}

File: page-object-model/src/main/java/pmf/imi/moodle/LoginPage.java
package pmf.imi.moodle;

import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.ui.ExpectedConditions;

public class LoginPage extends BasePageModel {

    public static final String LOGIN_URL =
            BASE_URL + "/login";


    @FindBy(name = "email")
    private WebElement emailInput;


    @FindBy(name = "password")
    private WebElement passwordInput;


    @FindBy(css = ".submit-button")
    private WebElement signInButton;


    @FindBy(css = ".error-message")
    private WebElement errorMessage;


    public LoginPage(WebDriver driver) {
        super(driver);
    }


    public void open() {

        driver.get(LOGIN_URL);

        wait.until(
                ExpectedConditions
                        .visibilityOf(emailInput)
        );
    }


    public void login(
            String email,
            String password) {

        wait.until(
                ExpectedConditions
                        .visibilityOf(emailInput)
        );

        emailInput.clear();
        emailInput.sendKeys(email);

        passwordInput.clear();
        passwordInput.sendKeys(password);

        signInButton.click();
    }


    public boolean isErrorDisplayed() {

        return wait.until(
                ExpectedConditions
                        .visibilityOf(errorMessage)
        ).isDisplayed();
    }


    public String getErrorMessage() {

        return wait.until(
                ExpectedConditions
                        .visibilityOf(errorMessage)
        ).getText();
    }


    public void waitForSuccessfulLogin() {

        wait.until(driver ->
                !driver.getCurrentUrl()
                        .contains("/login")
        );
    }
}